packages/mdc-ripple/_mixins.scss

```diff
@@ -21,68 +21,123 @@
-  --mdc-ripple-fg-translate-start: 0;
+@mixin mdc-ripple-common($query: mdc-feature-all()) {
+  $feat-animation: mdc-feature-create-target($query, animation);
+  $feat-structure: mdc-feature-create-target($query, structure);
 
-  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
-  will-change: transform, opacity;
+  // Ensure that styles needed by any component using MDC Ripple are emitted, but only once.
+  // (Every component using MDC Ripple imports these mixins, but doesn't necessarily import
+  // mdc-ripple.scss.)
+  @include mdc-feature-targets($feat-animation) {
+    @include mdc-base-emit-once("mdc-ripple/common/animation") {
+      @include mdc-ripple-keyframes_;
+    }
+  }
 
-  &::before,
-  &::after {
-    position: absolute;
-    border-radius: 50%;
-    opacity: 0;
-    pointer-events: none;
-    content: "";
+  @include mdc-feature-targets($feat-structure) {
+    @include mdc-base-emit-once("mdc-ripple/common/structure") {
+      // Styles used to detect buggy behavior of CSS custom properties in Edge.
+      // See: https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/11495448/
+      // This is included in _mixins.scss rather than mdc-ripple.scss so that it will be
+      // present for other components which rely on ripple as well as mdc-ripple itself.
+      .mdc-ripple-surface--test-edge-var-bug {
+        --mdc-ripple-surface-test-edge-var: 1px solid #000;
+
+        visibility: hidden;
+
+        &::before {
+          border: var(--mdc-ripple-surface-test-edge-var);
+        }
+      }
+    }
+  }
+}
+
+@mixin mdc-ripple-surface($query: mdc-feature-all()) {
+  $feat-animation: mdc-feature-create-target($query, animation);
+  $feat-structure: mdc-feature-create-target($query, structure);
+
+  @include mdc-feature-targets($feat-structure) {
+    --mdc-ripple-fg-size: 0;
+    --mdc-ripple-left: 0;
+    --mdc-ripple-top: 0;
+    --mdc-ripple-fg-scale: 1;
+    --mdc-ripple-fg-translate-end: 0;
+    --mdc-ripple-fg-translate-start: 0;
+
+    -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
+    will-change: transform, opacity;
+
+    &::before,
+    &::after {
+      position: absolute;
+      border-radius: 50%;
+      opacity: 0;
+      pointer-events: none;
+      content: "";
+    }
   }
 
   &::before {
-    // Also transition background-color to avoid unnatural color flashes when toggling activated/selected state
-    transition:
-      opacity $mdc-states-wash-duration linear,
-      background-color $mdc-states-wash-duration linear;
-    z-index: 1; // Ensure that the ripple wash for hover/focus states is displayed on top of positioned child elements
+    @include mdc-feature-targets($feat-animation) {
+      // Also transition background-color to avoid unnatural color flashes when toggling activated/selected state
+      transition:
+        opacity $mdc-states-wash-duration linear,
+        background-color $mdc-states-wash-duration linear;
+    }
+
+    @include mdc-feature-targets($feat-structure) {
```

Comment:
I think it is safe to merge these structure style to above feat-structure block.

Comment applies across this file.